File: 0020-Valid-Parentheses/solution.c
bool isValid(char* s) {
    if (s == NULL) {
        return false;
    }    

    int n = strlen(s);
    char stack[n];
    int top = -1;

    int i =0;

    while (s[i] != '\0') {
        if (s[i] == '[') {
            top++;
            stack[top] = ']';
        }
        else if (s[i] == '{') {
            top++;
            stack[top] = '}';
        }

        else if (s[i] == '(') {
            top++;
            stack[top] = ')';
        }

        else if(s[i] == ']' || s[i] == '}' || s[i] == ')') {
            if (top == -1) {
                return false;
            }
            else if (stack[top] != s[i]) {
                return false;
            }
            top--;
        }
        i++;
    }

    if (top != -1) {
        return false;
    }
    return true;
}

File: 0020-Valid-Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        if s is None:
            return False

        stack = []

        for c in s:
            if c == '[':
                stack.append(']')
            
            elif c == '{':
                stack.append('}')
            
            elif c == '(':
                stack.append(')')
            
            else:
                if not stack or stack.pop() != c:
                    return False

        
        return not stack
